examples/casestudies/inference.effekt.md

```diff
@@ -0,0 +1,617 @@
+      val density = 1.0 / (sqrt(variance) * (sqrt(2.0 * PI))) * exp((0.0 - 1.0 / 2.0) * (square(value - mean) / variance))
+      return density
+
+    case Uniform(lower, upper) =>
+      if (lower < value && value < upper) { 1.0 / (upper - lower) }
+      else { 0.0 }
+
+    case Beta(mean, sampleSize) =>
+      val alpha = mean * sampleSize
+      val beta = (1.0 - mean) * sampleSize
+      val density = (gamma(alpha + beta) / (gamma(alpha) * gamma(beta))) * pow(value, (alpha - 1.0)) * pow((1.0 - value), (beta - 1.0))
+      return density
+  }
+}
+```
+
+We define default handlers for all effects in this library
+
+```
+def handleSample[R] { program: () => R / sample } = {
+  try { program() }
+  with sample { dist => resume(draw(dist)) }
+}
+def handleObserve[R] { program: () => R / observe } = {
+  try { program() }
+  with observe { (value, dist) => do weight(density(value, dist)); resume(()) }
+}
+def handleRandom[R] { program: () => R / random }: R = {
+  try { program() }
+  with random { () => resume(random()) }
+}
+// pseudo random number generator
+def linearCongruentialGenerator[R](seed: Int) { prog: => R / random } : R = {
+  // parameters from Numerical Recipes (https://en.wikipedia.org/wiki/Linear_congruential_generator)
+  val a = 1664525
+  val c = 1013904223
+  val m = 1073741824
+  var x: Int = seed;
+  def next() = { x = mod((a * x + c), m); x }
+  try { prog() }
+  with random {
+    resume(next().toDouble / m.toDouble)
+  }
+}
+def handleWeight[R] { program: () => R / weight }: (R, Double)  = {
+  var current = 1.0
+  try { (program(), current) }
+  with weight { (prob) =>
+    current = current * prob;
+    resume(())
+  }
+}
+```
+
+With the effect `emit` it is also possible to limit possibly infinite loops. This allows for flexible numbers of steps to be performed with any algorithm that uses the effect `emit`.
+
+```
+def onEmit[A] { handler: A => Unit } { program: => Any / Emit[A] }: Unit = {
+  try { program(); () }
+  with Emit[A] {
+    def emit(element) = { handler(element); resume(()) }
+  }
+}
+
+def limit[A](n: Int) { program: => Any / Emit[A] }: Unit / Emit[A] = {
+  var steps = n
+  try { program(); () }
+  with Emit[A] {
+    def emit(element) = {
+      do emit(element);
+      steps = steps - 1;
+      if (steps > 0) { resume(()) }
+    }
+  }
+}
+
+def collect[A,R](n: Int){ program: () => R / Emit[A] }: List[A] = {
+  var result: List[A] = Nil()
+  var steps = n
+  try { program(); () }
+  with Emit[A] {
+    def emit(element) = {
+      result = Cons(element, result);
+      steps = steps - 1;
+      if (0 < steps) { resume(()) }
+    }
+  }
+  result.reverse
+}
+```
+
+## Rejection Sampling
+The effect `Weight` is also the basis of the rejection handling algorithm
+
+```
+def handleRejection[R] { program: () => R / weight }: R / random = {
+  try { program() }
+  with weight { (prob) =>
+    if (do random() < prob) { resume(()) }
+    else { handleRejection{ program } }
```

Comment:
```suggestion
    else { handleRejection {program} }
```
Actually in this case I was briefly confused. Maybe calling `program()` here is clearer.

Reply:
This cannot be done. I am not quite sure why, though.

```
/Users/david/Develop/effekt/inference-effekt/out/inference.js:1689
    return handleRejection_2899((weight$capability148573665550262348_20704) => {
                                ^

RangeError: Maximum call stack size exceeded
    at /Users/david/Develop/effekt/inference-effekt/out/inference.js:1689:33
    at /Users/david/Develop/effekt/inference-effekt/out/inference.js:1041:22
    at /Users/david/Develop/effekt/inference-effekt/out/inference.js:1041:22
    at /Users/david/Develop/effekt/inference-effekt/out/inference.js:1041:22
    at /Users/david/Develop/effekt/inference-effekt/out/inference.js:1041:22
    at /Users/david/Develop/effekt/inference-effekt/out/inference.js:1041:22
    at /Users/david/Develop/effekt/inference-effekt/out/inference.js:1041:22
    at /Users/david/Develop/effekt/inference-effekt/out/inference.js:1041:22
    at /Users/david/Develop/effekt/inference-effekt/out/inference.js:1041:22
    at /Users/david/Develop/effekt/inference-effekt/out/inference.js:1041:22

Node.js v22.5.1
[error] Process exited with non-zero exit code 1.
```